payroll/models/hr_payslip.py

```diff
@@ -575,44 +570,54 @@ def _compute_payslip_line(self, rule, localdict, lines_dict):
-            )
-            # set up localdict with current contract and employee values
-            localdict = dict(
-                baselocaldict, employee=contract.employee_id, contract=contract
-            )
-            for rule in self._get_salary_rules(contracts, payslip):
-                localdict["result"] = None
-                localdict["result_qty"] = 1.0
-                localdict["result_rate"] = 100
-                localdict["result_name"] = None
-                # check if the rule can be applied
-                if rule._satisfy_condition(localdict) and rule.id not in blacklist:
-                    localdict, lines_dict = payslip._compute_payslip_line(
-                        rule, localdict, lines_dict
-                    )
-                else:
-                    # blacklist this rule and its children
-                    blacklist += [id for id, seq in rule._recursive_search_of_rules()]
-            # call localdict_hook
-            localdict = self.localdict_hook(localdict, payslip)
-            # reset "current_contract" dict
-            baselocaldict["current_contract"] = {}
+        for payslip in self:
+            contracts = payslip._get_employee_contracts()
+            baselocaldict = payslip._get_baselocaldict(contracts)
+            for contract in contracts:
+                # assign "current_contract" dict
+                baselocaldict["current_contract"] = BrowsableObject(
+                    payslip.employee_id.id,
+                    payslip.get_current_contract_dict(contract, contracts),
+                    payslip.env,
+                )
+                # set up localdict with current contract and employee values
+                localdict = dict(
+                    baselocaldict, employee=contract.employee_id, contract=contract
+                )
+                for rule in payslip._get_salary_rules():
+                    localdict["result"] = None
+                    localdict["result_qty"] = 1.0
+                    localdict["result_rate"] = 100
+                    localdict["result_name"] = None
+                    # check if the rule can be applied
+                    if rule._satisfy_condition(localdict) and rule.id not in blacklist:
+                        localdict, _dict = payslip._compute_payslip_line(
+                            rule, localdict, lines_dict
+                        )
+                    else:
+                        # blacklist this rule and its children
+                        blacklist += [
+                            id for id, seq in rule._recursive_search_of_rules()
+                        ]
+                    lines_dict.update(_dict)
+                    # call localdict_hook
+                    localdict = payslip.localdict_hook(localdict)
+                # reset "current_contract" dict
+                baselocaldict["current_contract"] = {}
         return lines_dict
 
-    def localdict_hook(self, localdict, payslip):
```

Comment:
I think passing "payslip" here is actually useful because when you inherit the hook you would probably need some payslip value. Payslip will we available on self if I inherit this? If yes, we can remove it. If not, I need payslip object there. But maybe it's available on self.

Reply:
Yes, the payslip is available as self. You can access the payslip values as self.xxxx, self.yyy.

I was going to add a self.ensure_one() call here, but I wasn't sure how you would use it so I didn't add it.